File: kotlin/bc-envelope/src/main/kotlin/com/blockchaincommons/bcenvelope/EnvelopeException.kt
package com.blockchaincommons.bcenvelope

/**
 * Error types returned when operating on Gordian Envelopes.
 *
 * These errors capture various conditions that can occur when working with
 * envelopes, including structure validation, operation constraints, and
 * extension-specific errors.
 */
sealed class EnvelopeException(message: String) : Exception(message) {

    // -- Base Specification --

    /** Envelope was elided, so it cannot be compressed or encrypted. */
    class AlreadyElided : EnvelopeException("envelope was elided, so it cannot be compressed or encrypted")

    /** More than one assertion matches the predicate. */
    class AmbiguousPredicate : EnvelopeException("more than one assertion matches the predicate")

    /** Digest did not match. */
    class InvalidDigest : EnvelopeException("digest did not match")

    /** Invalid format. */
    class InvalidFormat : EnvelopeException("invalid format")

    /** A digest was expected but not found. */
    class MissingDigest : EnvelopeException("a digest was expected but not found")

    /** No assertion matches the predicate. */
    class NonexistentPredicate : EnvelopeException("no assertion matches the predicate")

    /** Cannot unwrap an envelope that was not wrapped. */
    class NotWrapped : EnvelopeException("cannot unwrap an envelope that was not wrapped")

    /** The envelope's subject is not a leaf. */
    class NotLeaf : EnvelopeException("the envelope's subject is not a leaf")

    /** The envelope's subject is not an assertion. */
    class NotAssertion : EnvelopeException("the envelope's subject is not an assertion")

    /** Assertion must be a map with exactly one element. */
    class InvalidAssertion : EnvelopeException("assertion must be a map with exactly one element")

    // -- Attachments Extension --

    /** Invalid attachment. */
    class InvalidAttachment : EnvelopeException("invalid attachment")

    /** Nonexistent attachment. */
    class NonexistentAttachment : EnvelopeException("nonexistent attachment")

    /** Ambiguous attachment. */
    class AmbiguousAttachment : EnvelopeException("ambiguous attachment")

    // -- Edges Extension --

    /** Edge missing 'isA' assertion. */
    class EdgeMissingIsA : EnvelopeException("edge missing 'isA' assertion")

    /** Edge missing 'source' assertion. */
    class EdgeMissingSource : EnvelopeException("edge missing 'source' assertion")

    /** Edge missing 'target' assertion. */
    class EdgeMissingTarget : EnvelopeException("edge missing 'target' assertion")

    /** Edge has duplicate 'isA' assertions. */
    class EdgeDuplicateIsA : EnvelopeException("edge has duplicate 'isA' assertions")

    /** Edge has duplicate 'source' assertions. */
    class EdgeDuplicateSource : EnvelopeException("edge has duplicate 'source' assertions")

    /** Edge has duplicate 'target' assertions. */
    class EdgeDuplicateTarget : EnvelopeException("edge has duplicate 'target' assertions")

    /** Edge has unexpected assertion. */
    class EdgeUnexpectedAssertion : EnvelopeException("edge has unexpected assertion")

    /** Nonexistent edge. */
    class NonexistentEdge : EnvelopeException("nonexistent edge")

    /** Ambiguous edge. */
    class AmbiguousEdge : EnvelopeException("ambiguous edge")

    // -- Compression Extension --

    /** Envelope was already compressed. */
    class AlreadyCompressed : EnvelopeException("envelope was already compressed")

    /** Cannot decompress an envelope that was not compressed. */
    class NotCompressed : EnvelopeException("cannot decompress an envelope that was not compressed")

    // -- Symmetric Encryption Extension --

    /** Envelope was already encrypted or compressed, so it cannot be encrypted. */
    class AlreadyEncrypted : EnvelopeException("envelope was already encrypted or compressed, so it cannot be encrypted")

    /** Cannot decrypt an envelope that was not encrypted. */
    class NotEncrypted : EnvelopeException("cannot decrypt an envelope that was not encrypted")

    // -- Known Values Extension --

    /** The envelope's subject is not a known value. */
    class NotKnownValue : EnvelopeException("the envelope's subject is not a known value")

    // -- Public Key Encryption Extension --

    /** Unknown recipient. */
    class UnknownRecipient : EnvelopeException("unknown recipient")

    // -- Encrypted Key Extension --

    /** Secret not found. */
    class UnknownSecret : EnvelopeException("secret not found")

    // -- Public Key Signing Extension --

    /** Could not verify a signature. */
    class UnverifiedSignature : EnvelopeException("could not verify a signature")

    /** Unexpected outer signature object type. */
    class InvalidOuterSignatureType : EnvelopeException("unexpected outer signature object type")

    /** Unexpected inner signature object type. */
    class InvalidInnerSignatureType : EnvelopeException("unexpected inner signature object type")

    /** Inner signature not made with same key as outer signature. */
    class UnverifiedInnerSignature : EnvelopeException("inner signature not made with same key as outer signature")

    /** Unexpected signature object type. */
    class InvalidSignatureType : EnvelopeException("unexpected signature object type")

    // -- SSKR Extension --

    /** Invalid SSKR shares. */
    class InvalidShares : EnvelopeException("invalid SSKR shares")

    // -- Types Extension --

    /** Invalid type. */
    class InvalidType : EnvelopeException("invalid type")

    /** Ambiguous type. */
    class AmbiguousType : EnvelopeException("ambiguous type")

    /** The subject of the envelope is not the unit value. */
    class SubjectNotUnit : EnvelopeException("the subject of the envelope is not the unit value")

    // -- Expressions Extension --

    /** Unexpected response ID. */
    class UnexpectedResponseID : EnvelopeException("unexpected response ID")

    /** Invalid response. */
    class InvalidResponse : EnvelopeException("invalid response")

    // -- Wrapped errors --

    /** SSKR error. */
    class SSKR(cause: Exception) : EnvelopeException("sskr error: ${cause.message}")

    /** CBOR error. */
    class Cbor(cause: Exception) : EnvelopeException("cbor error: ${cause.message}")

    /** Components error. */
    class Components(cause: Exception) : EnvelopeException("components error: ${cause.message}")

    /** General error. */
    class General(msg: String) : EnvelopeException("general error: $msg")

    companion object {
        /** Creates a general error with the given message. */
        fun msg(message: String): EnvelopeException = General(message)
    }
}

File: csharp/BCEnvelope/BCEnvelope/EnvelopeException.cs
using BlockchainCommons.BCComponents;
using BlockchainCommons.DCbor;

namespace BlockchainCommons.BCEnvelope;

/// <summary>
/// Base exception type for Gordian Envelope operations.
/// </summary>
public class EnvelopeException : Exception
{
    public EnvelopeException(string message) : base(message) { }
    public EnvelopeException(string message, Exception innerException) : base(message, innerException) { }

    // --- Base Specification ---

    /// <summary>
    /// Creates an exception for an already-elided envelope that cannot be compressed or encrypted.
    /// </summary>
    public static EnvelopeException AlreadyElided() =>
        new("envelope was elided, so it cannot be compressed or encrypted");

    /// <summary>
    /// Creates an exception when multiple assertions match a predicate.
    /// </summary>
    public static EnvelopeException AmbiguousPredicate() =>
        new("more than one assertion matches the predicate");

    /// <summary>
    /// Creates an exception for a digest validation failure.
    /// </summary>
    public static EnvelopeException InvalidDigest() =>
        new("digest did not match");

    /// <summary>
    /// Creates an exception for an invalid envelope format.
    /// </summary>
    public static EnvelopeException InvalidFormat() =>
        new("invalid format");

    /// <summary>
    /// Creates an exception when a digest is expected but not found.
    /// </summary>
    public static EnvelopeException MissingDigest() =>
        new("a digest was expected but not found");

    /// <summary>
    /// Creates an exception when no assertion matches a predicate.
    /// </summary>
    public static EnvelopeException NonexistentPredicate() =>
        new("no assertion matches the predicate");

    /// <summary>
    /// Creates an exception when unwrapping a non-wrapped envelope.
    /// </summary>
    public static EnvelopeException NotWrapped() =>
        new("cannot unwrap an envelope that was not wrapped");

    /// <summary>
    /// Creates an exception when the envelope subject is not a leaf.
    /// </summary>
    public static EnvelopeException NotLeaf() =>
        new("the envelope's subject is not a leaf");

    /// <summary>
    /// Creates an exception when the envelope subject is not an assertion.
    /// </summary>
    public static EnvelopeException NotAssertion() =>
        new("the envelope's subject is not an assertion");

    /// <summary>
    /// Creates an exception for an invalid assertion format.
    /// </summary>
    public static EnvelopeException InvalidAssertion() =>
        new("assertion must be a map with exactly one element");

    // --- Attachments Extension ---

    /// <summary>
    /// Creates an exception for an invalid attachment format.
    /// </summary>
    public static EnvelopeException InvalidAttachment() =>
        new("invalid attachment");

    /// <summary>
    /// Creates an exception for a nonexistent attachment.
    /// </summary>
    public static EnvelopeException NonexistentAttachment() =>
        new("nonexistent attachment");

    /// <summary>
    /// Creates an exception for an ambiguous attachment.
    /// </summary>
    public static EnvelopeException AmbiguousAttachment() =>
        new("ambiguous attachment");

    // --- Edges Extension ---

    /// <summary>
    /// Creates an exception when an edge is missing the 'isA' assertion.
    /// </summary>
    public static EnvelopeException EdgeMissingIsA() =>
        new("edge missing 'isA' assertion");

    /// <summary>
    /// Creates an exception when an edge is missing the 'source' assertion.
    /// </summary>
    public static EnvelopeException EdgeMissingSource() =>
        new("edge missing 'source' assertion");

    /// <summary>
    /// Creates an exception when an edge is missing the 'target' assertion.
    /// </summary>
    public static EnvelopeException EdgeMissingTarget() =>
        new("edge missing 'target' assertion");

    /// <summary>
    /// Creates an exception when an edge has duplicate 'isA' assertions.
    /// </summary>
    public static EnvelopeException EdgeDuplicateIsA() =>
        new("edge has duplicate 'isA' assertions");

    /// <summary>
    /// Creates an exception when an edge has duplicate 'source' assertions.
    /// </summary>
    public static EnvelopeException EdgeDuplicateSource() =>
        new("edge has duplicate 'source' assertions");

    /// <summary>
    /// Creates an exception when an edge has duplicate 'target' assertions.
    /// </summary>
    public static EnvelopeException EdgeDuplicateTarget() =>
        new("edge has duplicate 'target' assertions");

    /// <summary>
    /// Creates an exception when an edge has an unexpected assertion.
    /// </summary>
    public static EnvelopeException EdgeUnexpectedAssertion() =>
        new("edge has unexpected assertion");

    /// <summary>
    /// Creates an exception when a requested edge does not exist.
    /// </summary>
    public static EnvelopeException NonexistentEdge() =>
        new("nonexistent edge");

    /// <summary>
    /// Creates an exception when multiple edges match a query.
    /// </summary>
    public static EnvelopeException AmbiguousEdge() =>
        new("ambiguous edge");

    // --- Compression Extension ---

    /// <summary>
    /// Creates an exception for an already-compressed envelope.
    /// </summary>
    public static EnvelopeException AlreadyCompressed() =>
        new("envelope was already compressed");

    /// <summary>
    /// Creates an exception when decompressing a non-compressed envelope.
    /// </summary>
    public static EnvelopeException NotCompressed() =>
        new("cannot decompress an envelope that was not compressed");

    // --- Symmetric Encryption Extension ---

    /// <summary>
    /// Creates an exception for an already-encrypted or compressed envelope.
    /// </summary>
    public static EnvelopeException AlreadyEncrypted() =>
        new("envelope was already encrypted or compressed, so it cannot be encrypted");

    /// <summary>
    /// Creates an exception when decrypting a non-encrypted envelope.
    /// </summary>
    public static EnvelopeException NotEncrypted() =>
        new("cannot decrypt an envelope that was not encrypted");

    // --- Known Values Extension ---

    /// <summary>
    /// Creates an exception when the envelope subject is not a known value.
    /// </summary>
    public static EnvelopeException NotKnownValue() =>
        new("the envelope's subject is not a known value");

    /// <summary>
    /// Creates an exception when the subject of the envelope is not the unit value.
    /// </summary>
    public static EnvelopeException SubjectNotUnit() =>
        new("the subject of the envelope is not the unit value");

    // --- Public Key Encryption Extension ---

    /// <summary>
    /// Creates an exception for an unknown recipient.
    /// </summary>
    public static EnvelopeException UnknownRecipient() =>
        new("unknown recipient");

    // --- Encrypted Key Extension ---

    /// <summary>
    /// Creates an exception for an unknown secret.
    /// </summary>
    public static EnvelopeException UnknownSecret() =>
        new("secret not found");

    // --- Public Key Signing Extension ---

    /// <summary>
    /// Creates an exception for a signature verification failure.
    /// </summary>
    public static EnvelopeException UnverifiedSignature() =>
        new("could not verify a signature");

    /// <summary>
    /// Creates an exception for an invalid outer signature object type.
    /// </summary>
    public static EnvelopeException InvalidOuterSignatureType() =>
        new("unexpected outer signature object type");

    /// <summary>
    /// Creates an exception for an invalid inner signature object type.
    /// </summary>
    public static EnvelopeException InvalidInnerSignatureType() =>
        new("unexpected inner signature object type");

    /// <summary>
    /// Creates an exception when the inner signature is not made with the same key.
    /// </summary>
    public static EnvelopeException UnverifiedInnerSignature() =>
        new("inner signature not made with same key as outer signature");

    /// <summary>
    /// Creates an exception for an unexpected signature object type.
    /// </summary>
    public static EnvelopeException InvalidSignatureType() =>
        new("unexpected signature object type");

    // --- SSKR Extension ---

    /// <summary>
    /// Creates an exception for invalid SSKR shares.
    /// </summary>
    public static EnvelopeException InvalidShares() =>
        new("invalid SSKR shares");

    // --- Types Extension ---

    /// <summary>
    /// Creates an exception for an invalid type.
    /// </summary>
    public static EnvelopeException InvalidType() =>
        new("invalid type");

    /// <summary>
    /// Creates an exception for ambiguous type information.
    /// </summary>
    public static EnvelopeException AmbiguousType() =>
        new("ambiguous type");

    // --- Expressions Extension ---

    /// <summary>
    /// Creates an exception for an unexpected response ID.
    /// </summary>
    public static EnvelopeException UnexpectedResponseID() =>
        new("unexpected response ID");

    /// <summary>
    /// Creates an exception for an invalid response.
    /// </summary>
    public static EnvelopeException InvalidResponse() =>
        new("invalid response");

    // --- Wrapped Errors ---

    /// <summary>
    /// Creates an exception wrapping a CBOR exception.
    /// </summary>
    public static EnvelopeException FromCborException(CborException inner) =>
        new($"dcbor error: {inner.Message}", inner);

    /// <summary>
    /// Creates an exception wrapping a BCComponents exception.
    /// </summary>
    public static EnvelopeException FromComponentsException(BCComponentsException inner) =>
        new($"components error: {inner.Message}", inner);

    /// <summary>
    /// Creates an exception wrapping an SSKR exception.
    /// </summary>
    public static EnvelopeException FromSskrException(Exception inner) =>
        new($"sskr error: {inner.Message}", inner);

    /// <summary>
    /// Creates a general envelope exception with a custom message.
    /// </summary>
    public static EnvelopeException General(string message) =>
        new($"general error: {message}");
}
